src/mqtt.cpp

```diff
@@ -289,6 +290,7 @@ mqtt_connect()
     event_send(doc);
 
     // Publish MQTT override/claim
+    mqtt_publish_config();
```

Comment:
you should then not publish it at connect and wait for the loop to compare the config_version, as when the config will be loaded, it will increment it so this shouldn't be long to wait for ;)
And I bet the data will be ready then

Reply:
Ok so I confirm the behaviour. There is no update and the `mqtt_loop` is not working. So I have updated the code also for the schedule, claims, override and config. Because if your comments holds true for the config, it should also hold true for the rest right ? So no need to publish them once connected since they should be published in the loop. And to make sure of that, I've changed them to `-1` because the initial value of version is `0`.

Reply:
![image](https://github.com/OpenEVSE/ESP32_WiFi_V4.x/assets/61346/6908f9b7-0540-4d41-aa52-03e0a3736199)

So I will leave the config publish once connected, like it is done for the claims, override and schedule...

Reply:
it was to solve the fact config was not ready when you first publish it. 
If it's not incremented, let's just do it at config load.

Reply:
> it was to solve the fact config data were not totally ready when you first publish it.
> If it's not incremented yet, let's just do it at config load, and don't publish when version is still 0

But `config_version` is ONLY incremented after a config change happens from the UI.

So I don't get the _don't publish when version is still 0_... Version WILL be 0.

Reply:
I've wrote: 
>If it's not incremented yet, let's just do it at config load

was meant to add an increment at config load when the wifi fw initialise. I know it is not yet

Reply:
The problem I saw with the EVSE data not being present in the config json was when I flipped the electtic breaker to shutdown and restart my electrical box.

When doing that, both the evse card + the gateway were rebooted. And at this point, the publication of the config does not have the evse data.

This is not happening when we just restart of flash the gateway.

Reply:
but I think just do in the mqtt loop() the first /config publish and not in begin() should solve the uninitalised values published.
use -1 as default configVersion, check if -1 in the loop, publish config then set version to 0.

Reply:
> I think you can just do in the mqtt loop() the first /config publish and not in begin() should solve the uninitalised values published.
> use -1 as default configVersion, check if -1 in the loop, publish config then set configVersion to 0, it will match config_version then

tried that in the loop:

```json
    if(configVersion == -1) {
      mqtt_publish_config();
      configVersion = 0;
    }
```

but this is not working at all... Nothing gets published.

Reply:
So either we keep it as-is (config exposed once mqtt is connected) and if the box gets some power and evse is not yet started, then we have to restart the gateway, or we have to call `mqtt_publish_config` from where the evse gets ready to be sure...

Reply:
you better find why it's not publishing in your test above than find a hack to solve the issue. ;)

Reply:
> 

I've not a C guy... What options do I have to debug that ? I've tried the debugvar and some logging but do not see anything, which suggests the loop is not running maybe ? Or could it be the assignment of -1 to an unsigned int that is preventing the comparison ?

Also, what is ensuring that the evse will be ready at the first run of the loop ?

Isn't it more reliable to trigger the config publish once the evse is ready since the config system now depends on teh evse settings to be available ?

Reply:
@KipK : I've implemented the listener :

https://github.com/OpenEVSE/ESP32_WiFi_V4.x/compare/29c5a9d772dd082969269594c85e95228bacf04a..f6c5d54ba5a81471912512069357ab22a7f4bf79

So with all that it works perfectly:

- if mqtt connects before evse boots, the event sent from evse manager will take care of sending the right config update just after the previous one
- if evse boots before mqtt connects, the initial send of the config from mqtt will take care of the update (the one coming from evse won't happen because mqtt not connected)

I tested a complete reload of the box (electricity off/on) - it works.

Today i've flashed the openevse firmware 3 times to test some flags and each time I was turning power on, the config was badly published - with all evse settings missing (when published after mqtt connects because evse was not yet available). 

So this is really a corner case  that we are trying to handle here...

Reply:
I was thinking for another PR to add a variable for this, could be usefull yes.
But I don't think there's need to publish it, it's used for internal functions only for a fraction of a second. 
Perhaps an isEvseReady() function cascading in evseManager, then evseMonitor ( with a private _evse_ready flag set to true at the end of the evse ready callback ), should be more consistant with the codebase.

Reply:
but mostly if you check the evse state to be something else than OPENEVSE_STATE_STARTING, means it is ready

Reply:
OK! So I in summary I will add a boolean flag, set to true at `OPENEVSE_STATE_STARTING`, and I will expose it through the different apis, and access its getter in the new `mqtt_publush_config` method, to only expose if evse has started.
- I can keep the publish call at mqtt connection time, like it is done for the claims, override, schedule (because most of the time evse should be started at that point)
- I will track in the loop this initial publish if it has to be done also if the initial mqtt_publich_config call was not able to do it

Reply:
@KipK :  I think we overthink it ;-)

1. We just have to expose a flag in the evse to know if is has started or not.
2. In app config, when we add the evse config, we check this flag and if evse didn't start yet, we add instead `evse_starting = true`. If evse has started, we add the settings but not the boolean.

So this will work because any client receiving the partial config will be able to know if it is complete or not.

Besides, Home Assistant has a way to tell whether a component is available or not by looking at the content of the payload. So if I see `evse_starting = true`, then I can tell HA that the sensor is not yet available.

in the mqtt:
- we punlish at startup
-in the loop, we then also check this status and we publish again once evse is started. Then after only on config change,

Reply:
We're talking about evse data from the module that are not part of the real config anyway and are not settable through this endpoint.   The real config is always ready anyway.  On UI http side it grabs the config after receiving /status and other endpoints so all data are always ready at this stage.  (  the time it takes to refresh the UI the evse has already booted also ).

But I guess we could introduce a config_version change when those evse data are ready at the end of the evse ready callback.? 

@jeremypoulter how would you handle this ?

Reply:
Agree: Introducing a `config_version ` change so that the loop publishes would solve it yes, and remove the need to publish an eventually partial config when mqtt connects/

But can (should we) access this `config_version` variable from the evse ready clalback ?

Reply:
@KipK I have implemented that and it works perfectly, config is incremented then pushed in mqtt
![image](https://github.com/OpenEVSE/ESP32_WiFi_V4.x/assets/61346/7883ec23-1495-44ea-aee7-facd589b07ba)

Reply:
Ahah! So after testing, it works perfectly when the whole box is powered downed and up again. 
But if the gateway is restarted alone and the mqtt broker does not have the config topic, it won't be created because evse has already booted. So we need also need to publish at mqtt connection.

Reply:
PR rebased and updated.

Reply:
> downed and up again.
> But if the gateway is restarted alone and the mqtt broker does not have the config topic, it won't be created because evse has already booted. So we need also need to publish at mqtt connection.

if the evse is restarting alone, then it will call increment_config_version() from the EvseMonitor::evseBoot() , so you should get the config_version change in the mqtt loop too and publish it as we said above.

Reply:
Perhaps you can only publish at mqtt connection if the  evse state is != from OPENEVSE_STATE_STARTING  ?

Reply:
> Perhaps you can only publish at mqtt connection if the evse state is != from OPENEVSE_STATE_STARTING ?

For the rare case when it will happen... Is it worth coupling the mqtt layer with the evse state like that ?

Reply:
> Mqtt already use EvseManager, you can call it to get the current state with evse getEvseState()

Oh right I didn't notice the transitive includes....

So I've updated like that instead:

```c
void
mqtt_publish_config() {
  if(!config_mqtt_enabled() || !mqttclient.connected() || evse.getEvseState() == OPENEVSE_STATE_STARTING) {
    return;
  }
```